From ccba03d4fc4f8fe587c198e5dc4dec70b55109ae Mon Sep 17 00:00:00 2001
From: LuoYi <langisme@qq.com>
Date: Wed, 12 Jan 2022 12:04:17 +0800
Subject: [PATCH] Hide blur actor when switching workspace

on_visiable_changed() callback has been used to hide blur actor when
switching window between workspace, so I have to add it back.
---
 src/compositor/meta-window-actor.c | 27 ++++++++++++++++++++++-----
 1 file changed, 22 insertions(+), 5 deletions(-)

diff --git a/src/compositor/meta-window-actor.c b/src/compositor/meta-window-actor.c
index 5d6cdcc627..9d2994e74e 100644
--- a/src/compositor/meta-window-actor.c
+++ b/src/compositor/meta-window-actor.c
@@ -65,6 +65,7 @@ typedef struct _MetaWindowActorPrivate
   ClutterActor *blur_actor;
   MetaShellBlurEffect *blur_effect;
 
+  ulong visible_changed_id;
   ulong wm_class_changed_id;
   int geometry_scale;
 
@@ -661,6 +662,20 @@ init_surface_actor (MetaWindowActor *self)
     meta_window_actor_assign_surface_actor (self, surface_actor);
 }
 
+static void
+on_visible_changed (MetaWindowActor *self)
+{
+  MetaWindowActorPrivate *priv = meta_window_actor_get_instance_private (self);
+
+  if (!priv->blur_actor)
+    return;
+
+  if (priv->visible && !meta_window_actor_effect_in_progress (self))
+    clutter_actor_show(priv->blur_actor);
+  else
+    clutter_actor_hide(priv->blur_actor);
+}
+
 static void
 on_wm_class_changed (MetaWindow *self,
                      gpointer    user_data)
@@ -704,6 +719,9 @@ meta_window_actor_constructed (GObject *object)
     priv->wm_class_changed_id =
       g_signal_connect (window, "notify::wm-class", G_CALLBACK (on_wm_class_changed), object);
   }
+
+  priv->visible_changed_id = 
+    g_signal_connect (object, "notify::visible", G_CALLBACK (on_visible_changed), NULL);
 }
 
 static void
@@ -722,6 +740,7 @@ meta_window_actor_dispose (GObject *object)
 
   priv->disposed = TRUE;
 
+  g_clear_signal_handler(&priv->visible_changed_id, object);
   meta_compositor_remove_window_actor (compositor, self);
 
   g_clear_object (&priv->window);
@@ -869,6 +888,7 @@ meta_window_actor_effect_in_progress (MetaWindowActor *self)
     meta_window_actor_get_instance_private (self);
 
   return (priv->minimize_in_progress ||
+          priv->unminimize_in_progress ||
           priv->size_change_in_progress ||
           priv->map_in_progress ||
           priv->destroy_in_progress);
@@ -922,7 +942,7 @@ meta_window_actor_show_blur (MetaWindowActor *self)
   if (!priv->blur_actor)
     return;
 
-  if (priv->visible && !priv->unminimize_in_progress)
+  if (priv->visible && !meta_window_actor_effect_in_progress (self))
     clutter_actor_show(priv->blur_actor);
   else
     clutter_actor_hide(priv->blur_actor);
@@ -1022,10 +1042,7 @@ meta_window_actor_after_effects (MetaWindowActor *self)
       g_signal_emit (self, signals[EFFECTS_COMPLETED], 0);
       meta_window_actor_sync_visibility (self);
       meta_window_actor_sync_actor_geometry (self, FALSE);
-      if (priv->round_clip_effect && !priv->unminimize_in_progress)
-        {
-         meta_window_actor_show_blur(self);
-        }
+      meta_window_actor_show_blur(self);
     }
 
   clutter_stage_repick_device (stage, clutter_seat_get_pointer (seat));

